File: semana_6_patrones_estructurales/desarrollo/desarrollo1_adapter_pagos.py
"""
DESARROLLO 1: Adaptador de pagos (dolares a soles)
(EJERCICIO NUEVO - NO es el ejemplo del PDF)
"""


class PasarelaDolares:
    def pagar_usd(self, monto):
        return f"Pago: ${monto} USD"


class AdaptadorSoles:
    def __init__(self, pasarela):
        self.pasarela = pasarela
        self.tipo_cambio = 3.75
    
    def pagar_soles(self, monto):
        monto_usd = monto / self.tipo_cambio
        return f"Convertido: S/{monto} -> {self.pasarela.pagar_usd(monto_usd)}"


pasarela = PasarelaDolares()
adaptador = AdaptadorSoles(pasarela)
print(adaptador.pagar_soles(375))

File: semana_6_patrones_estructurales/desarrollo/desarrollo1_adapter_pagos.cpp
/*
 * DESARROLLO 1: Adaptador de pagos (dolares a soles)
 * (EJERCICIO NUEVO - NO es el ejemplo del PDF)
 */

#include <iostream>
#include <string>

using namespace std;

class PasarelaDolares {
public:
    string pagarUSD(double monto) {
        return "Pago: $" + to_string(monto) + " USD";
    }
};

class AdaptadorSoles {
private:
    PasarelaDolares* pasarela;
    double tipoCambio = 3.75;
    
public:
    AdaptadorSoles(PasarelaDolares* p) : pasarela(p) {}
    
    string pagarSoles(double monto) {
        double montoUSD = monto / tipoCambio;
        return "Convertido: S/" + to_string(monto) + " -> " + pasarela->pagarUSD(montoUSD);
    }
};

int main() {
    PasarelaDolares pasarela;
    AdaptadorSoles adaptador(&pasarela);
    cout << adaptador.pagarSoles(375) << endl;
    return 0;
}
